File: LeetCode/Medium/0015. 3Sum.cpp
class Solution {
public:
    
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res; 
        int idx1, idx2;
        if(nums.size() < 3) return res;
        
        sort(nums.begin(), nums.end());
        
        for(int i = 0; i < nums.size()-2; i++) {
            if(i != 0 && nums[i] == nums[i-1]) continue;
            idx1 = i + 1; idx2 = nums.size()-1;
            
            while(idx1 < idx2) {
                if(idx1 != i+1 && nums[idx1] == nums[idx1-1]) {
                    idx1++; continue;
                } else if(idx2 != nums.size()-1 && nums[idx2] == nums[idx2+1]) {
                    idx2--; continue;
                }
                
                int sum = nums[i] + nums[idx1] + nums[idx2];
                
                if(sum == 0) {
                    vector<int> triplet = {nums[i], nums[idx1], nums[idx2]};
                    res.push_back(triplet);
                    idx2--;
                } else if(sum < 0) {
                    idx1++;
                } else if(sum > 0) {
                    idx2--;
                }
            }
        }
        
        return res;
        
    }
};

File: LeetCode/Medium/0015. 3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]

        algo: 
        1. sort nums
        2. start from 1st elem - use 2 pointers to find the rest of 2 elem
        3. repeat step2 for all elem
        """

        ans = []
        nums.sort()

        for i in range(len(nums)):
            if i >= 1 and nums[i] == nums[i-1]: continue

            start, end = i+1, len(nums)-1
            while start < end:
                cur_sum = nums[i] + nums[start] + nums[end]
                if cur_sum == 0:
                    ans.append([nums[i], nums[start], nums[end]])
                    start += 1
                    while start < len(nums) and nums[start] == nums[start-1]:
                        start += 1
                    end -= 1
                elif cur_sum > 0:
                    end -= 1
                elif cur_sum < 0:
                    start += 1
        
        return ans
            

